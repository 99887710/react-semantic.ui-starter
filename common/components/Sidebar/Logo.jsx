import React, {Component} from 'react';
import {Image} from 'semantic-ui-react'

class Logo extends Component {
    constructor(props) {
        super(props)
    }

    shouldComponentUpdate() {
        return false
    }

    render() {
        let url = 'https://raw.githubusercontent.com/reactjs/redux/master/logo/logo.png'
        return (
            <div className="logo">
                <Image src={url} centered height="34px"/>
            </div>
        )
    }
}

export default Logo
