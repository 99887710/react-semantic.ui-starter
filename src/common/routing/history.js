// @flow
import {createBrowserHistory, createMemoryHistory} from 'history'
const selectHistory = (): Function => {
	return process.env.BROWSER ? createBrowserHistory : createMemoryHistory
}

export default selectHistory()()
