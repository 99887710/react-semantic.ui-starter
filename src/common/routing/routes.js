// @flow
import {Route} from 'react-router-dom'
import RouteAuth from 'components/addons/RouteAuth'
// import Dashboard from 'containers/Dashboard'
// import Login from 'containers/Login'
import type {RouteItem} from 'types'

const loadLazyComponentFnConstructor = (url: string) => async () => {
	// NOTE: there isn't any duplication here
	// Read Webpack docs about code-splitting for more info.
	if (process.env.BROWSER) {
		const loadComponent = await import(/* webpackMode: "lazy", webpackChunkName: "[request].lazy" */ `containers/${url}/index.jsx`)
		return loadComponent
	}
	const loadComponent = await import(/* webpackMode: "eager" */ `containers/${url}/index.jsx`)
	return loadComponent
}

// FIXME: sidebar routes and app routes should be separated!
const routes: Array<RouteItem> = [
	{
		path: '/',
		exact: true,
		icon: 'newspaper',
		name: 'Dashboard',
		sidebarVisible: true,
		lazy: true,
		tag: RouteAuth,
		component: loadLazyComponentFnConstructor('Dashboard')
	},
	{
		path: '/links',
		name: 'Links',
		exact: true,
		lazy: true,
		icon: 'bookmark',
		sidebarVisible: true,
		tag: RouteAuth,
		component: loadLazyComponentFnConstructor('Links')
	},
	{
		external: true,
		path: 'https://github.com/Metnew/react-semantic.ui-starter',
		icon: 'github',
		name: 'Github',
		sidebarVisible: true
	},
	{
		path: '/auth',
		name: 'Auth',
		lazy: true,
		tag: Route,
		component: loadLazyComponentFnConstructor('Login')
	}
]

export default routes
