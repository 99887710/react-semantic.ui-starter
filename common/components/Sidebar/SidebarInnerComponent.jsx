import React, {Component} from 'react'
import PropTypes from 'prop-types'
import {Link} from 'react-router-dom'
import {Menu, Icon} from 'semantic-ui-react'
import Logo from './Logo'
import './Sidebar.scss'

export default class SidebarInnerComponent extends Component {
    constructor(props) {
        super(props);
    }

    static propTypes = {
        logout: PropTypes.func,
        routing: PropTypes.array
    }

    shouldComponentUpdate() {
        return false
    }

    render() {
        const {logout, routing} = this.props

        let routes = routing.map((route, i) => {
            let {external, href, icon, name} = route
            let propsMenuItem = {
                as: external ? 'a' : Link,
                link: true,
                key: i,
                [external ? 'href' : 'to']: href
            }

            return (
                <Menu.Item {...propsMenuItem} icon>
                    <Icon name={icon}/> {name}
                </Menu.Item>
            )
        })



        return (
            <div>
                <Logo centered/>
                {routes}
                <Menu.Item className="logout" onClick={logout}>
                    <Icon name='sign out'/>
                    Logout
                </Menu.Item>
            </div>
        )
    }
}
