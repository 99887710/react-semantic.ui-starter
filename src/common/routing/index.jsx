/**
 * @flow
 */
import React from 'react'
import {Route} from 'react-router-dom'
import RouteAuth from 'components/addons/RouteAuth'
import {createBrowserHistory, createMemoryHistory} from 'history'
import type {RouteItem} from 'types'

const getHistory = () => {
	if (process.env.BROWSER) {
		return createMemoryHistory()
	}
	return createBrowserHistory()
}

export const history = getHistory()

const loadLazyComponent = (url: string) => async () => {
	// NOTE: there isn't any duplication here
	// Read Webpack docs about code-splitting for more info.
	if (process.env.BROWSER) {
		const loadComponent = await import(/* webpackMode: "lazy-once", webpackChunkName: "lazy-containers" */ `containers/${url}/index.jsx`)
		return loadComponent
	}
	const loadComponent = await import(/* webpackMode: "eager", webpackChunkName: "lazy-containers" */ `containers/${url}/index.jsx`)
	return loadComponent
}

export const routes: Array<RouteItem> = [
	{
		path: '/',
		exact: true,
		lazy: true,
		icon: 'newspaper',
		name: 'Dashboard',
		sidebarVisible: true,
		tag: RouteAuth,
		component: loadLazyComponent('Dashboard')
	},
	{
		path: '/users',
		name: 'Users',
		exact: true,
		lazy: true,
		icon: 'users',
		sidebarVisible: true,
		tag: RouteAuth,
		component: loadLazyComponent('Users')
	},
	{
		external: true,
		path: 'https://github.com/Metnew/react-semantic.ui-starter',
		icon: 'github',
		name: 'Github',
		sidebarVisible: true
	},
	{
		path: '/auth',
		name: 'Auth',
		lazy: true,
		tag: Route,
		component: loadLazyComponent('Login')
	},
	{
		path: '/users/:id',
		name: 'User',
		lazy: true,
		exact: true,
		tag: RouteAuth,
		component: loadLazyComponent('UserItem')
	}
]
