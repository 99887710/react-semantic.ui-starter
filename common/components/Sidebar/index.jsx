import React, {Component} from 'react'
import PropTypes from 'prop-types'
import {Menu, Sidebar} from 'semantic-ui-react'
import SidebarInnerComponent from './SidebarInnerComponent'
import _ from 'lodash'
import './Sidebar.scss'

export default class SidebarComponent extends Component {
    constructor(props) {
        super(props);
    }

    static propTypes = {
        open: PropTypes.bool,
        logout: PropTypes.func,
        routing: PropTypes.array,
        isMobile: PropTypes.bool
    }

    shouldComponentUpdate(nextProps) {
        // re-render on isMobile changed
        if (nextProps.isMobile !== this.props.isMobile) {
            return true
        }
        // maybe routing isn't the same
        let routingSame = !_.isEqual(nextProps.routing, this.props.routing)
        return routingSame
    }

    render() {
        const {open, logout, routing, isMobile} = this.props

        let sidebarProps = {
            visible: open || !isMobile,
            as: Menu,
            vertical: true,
            icon: 'labeled',
            animation: 'push',
            width: 'thin'
        }

        let sidebarInnerComponentProps = {
            logout,
            routing
        }

        return (
            <Sidebar {...sidebarProps}>
                <SidebarInnerComponent {...sidebarInnerComponentProps} />
            </Sidebar>
        )
    }
}
