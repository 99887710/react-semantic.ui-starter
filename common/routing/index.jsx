import React from 'react';
import {Route, Redirect, Switch} from 'react-router';
import {createBrowserHistory} from 'history'
import {App, Inbox, Dashboard, Login} from 'containers';
import {RouteAuth} from 'components'

export const history = getHistory()

export const appRouting = [
    {
        path: '/',
        icon: 'newspaper',
        name: 'Dashboard',
        exact: true,
        sidebarVisible: true,
        tag: RouteAuth,
        component: Dashboard
    }, {
        path: '/inbox',
        name: 'Inbox',
        icon: 'comments outline',
        sidebarVisible: true,
        tag: RouteAuth,
        component: Inbox
    }, {
        external: true,
        path: 'https://github.com/Metnew/react-semantic.ui-starter',
        icon: 'github',
        name: 'Github',
        sidebarVisible: true
    }, {
        path: '/auth',
        name: 'Auth',
        tag: Route,
        component: Login
    }
]

/**
 * Returns application routing with protected by AuthCheck func routes
 * @param {Function} AuthCheck checks is user logged in
 */
export const Routing = (AuthCheck) => {
    // remove components that aren't application routes, (e.g. github link in sidebar)
    let routes = appRouting.filter(a => a.tag || a.component).map(a => {
        let {tag, path, exact, component} = a
        let b = {
            tag,
            path,
            exact,
            component
        }
        return b
    })
    // render components that are inside Switch (main view)
    let routesRendered = routes.map((a, i) => {
        // get tag for Route. is it RouteAuth `protected route` or Route?
        let Tag = a.tag
        delete a.tag
        return (<Tag key={Math.random()} {...a}/>)
    })

    return (
        <App>
            <Switch>
                {rot}
                <Redirect from="*" to="/"/>
            </Switch>
        </App>
    )
}

export const sidebarRouting = [
    {
        name: 'Dashboard',
        href: '/',
        icon: 'newspaper'
    }, {
        name: 'Inbox',
        href: '/inbox',
        icon: 'comments outline'
    }, {
        name: 'Github',
        external: true,
        href: 'https://github.com/Metnew/react-semantic.ui-starter',
        icon: 'github'
    }
]

function getHistory() {
    const basename = process.env.BUILD_DEMO
        ? '/react-semantic.ui-starter'
        : ''

    return createBrowserHistory({basename})
}
