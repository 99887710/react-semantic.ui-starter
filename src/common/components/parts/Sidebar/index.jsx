// @flow
import React, {Component} from 'react'
import {connect} from 'react-redux'
import {Menu, Icon} from 'semantic-ui-react'
import {NavLink} from 'react-router-dom'
import {LOGOUT_AUTH} from 'actions/auth'
import {getSidebarRoutes} from 'routing'
import {getLayoutState} from 'selectors'
import {
	StyledSidebar,
	SidebarLogo,
	SidebarLogoContainer,
	SidebarItem,
	SidebarLogoutItem
} from './style'
import {Spacer} from 'styles/base'
import type {RouteItem} from 'types'

type Props = {
	open: boolean,
	logout: () => void,
	routing: Array<RouteItem>,
	isMobile: boolean
}

class SidebarComponent extends Component <Props> {
	props: Props

	render () {
		const {open, logout, routing, isMobile} = this.props

		const sidebarProps = {
			visible: open || !isMobile,
			as: Menu,
			vertical: true,
			icon: 'labeled',
			animation: 'push',
			width: 'thin'
		}

		const routes = routing.map((route, i) => {
			const {external, path, strict, exact, meta} = route
			const {icon, name} = meta
			// Props that are common for both "<a>" and "RR Link"
			const basicProps = {
				as: external ? 'a' : NavLink,
				link: true,
				[external ? 'href' : 'to']: path
			}

			// Is it's RR Link, then it needs additional props
			const externalProps = external
				? {}
				: {
					strict,
					exact,
					activeClassName: 'active'
				}

			// Summarize
			const propsMenuItem = {
				...externalProps,
				...basicProps
			}

			return (
				<SidebarItem key={i} {...propsMenuItem} icon>
					<Icon name={icon} /> {name}
				</SidebarItem>
			)
		})

		return (
			<StyledSidebar {...sidebarProps}>
				<SidebarLogoContainer href="https://github.com/Metnew/suicrux">
					<SidebarLogo alt="logo" shape="circular" centered />
				</SidebarLogoContainer>
				{routes}
				<Spacer />
				<SidebarLogoutItem onClick={logout}>
					<Icon name="sign out" />
					Logout
				</SidebarLogoutItem>
			</StyledSidebar>
		)
	}
}

const mapStateToProps = (state) => {
	const {isMobile, sidebarOpened} = getLayoutState(state)
	const routing = getSidebarRoutes()

	return {
		routing,
		open: sidebarOpened,
		isMobile
	}
}

const mapDispatchToProps = (dispatch) => ({
	logout () {
		dispatch(LOGOUT_AUTH())
	}
})

export default connect(mapStateToProps, mapDispatchToProps)(SidebarComponent)
