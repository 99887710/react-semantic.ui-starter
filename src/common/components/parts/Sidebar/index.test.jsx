import React from 'react'
import Sidebar from 'components/parts/Sidebar'
import {shallow} from 'enzyme'
import toJson from 'enzyme-to-json'
import {routes as routing} from 'routing'

const props = {
	isMobile: true,
	logout: () => ({}),
	routing,
	open: true
}

describe('Sidebar component', () => {
	it('same snapshot if Sidebar is open and it is on mobile screen', () => {
		const component = shallow(<Sidebar {...props} />)
		expect(toJson(component)).toMatchSnapshot()
	})
})
