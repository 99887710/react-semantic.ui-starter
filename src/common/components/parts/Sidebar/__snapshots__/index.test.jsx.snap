// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar component same snapshot if Sidebar is open and it is on mobile screen 1`] = `
<Styled(Sidebar)
  animation="push"
  as={[Function]}
  icon="labeled"
  vertical={true}
  visible={true}
  width="thin"
>
  <styled.a
    href="https://github.com/Metnew/react-semantic.ui-starter"
  >
    <Styled(Logo)
      alt="logo"
      centered={true}
      shape="circular"
    />
  </styled.a>
  <Styled(MenuItem)
    activeClassName="active"
    as={[Function]}
    exact={true}
    icon={true}
    link={true}
    to="/"
  >
    <Icon
      as="i"
      name="newspaper"
    />
     
    Dashboard
  </Styled(MenuItem)>
  <Styled(MenuItem)
    activeClassName="active"
    as={[Function]}
    exact={true}
    icon={true}
    link={true}
    to="/users"
  >
    <Icon
      as="i"
      name="users"
    />
     
    Users
  </Styled(MenuItem)>
  <Styled(MenuItem)
    as="a"
    href="https://github.com/Metnew/react-semantic.ui-starter"
    icon={true}
    link={true}
  >
    <Icon
      as="i"
      name="github"
    />
     
    Github
  </Styled(MenuItem)>
  <Styled(MenuItem)
    activeClassName="active"
    as={[Function]}
    icon={true}
    link={true}
    to="/auth"
  >
    <Icon
      as="i"
    />
     
    Auth
  </Styled(MenuItem)>
  <Styled(MenuItem)
    activeClassName="active"
    as={[Function]}
    exact={true}
    icon={true}
    link={true}
    to="/users/:id"
  >
    <Icon
      as="i"
    />
     
    User
  </Styled(MenuItem)>
  <styled.span />
  <Styled(MenuItem)
    onClick={[Function]}
  >
    <Icon
      as="i"
      name="sign out"
    />
    Logout
  </Styled(MenuItem)>
</Styled(Sidebar)>
`;
